themes/default/content/blog/shared-config-with-aws-systems-manager-parameter-store/index.md

```diff
@@ -0,0 +1,405 @@
+$ mkdir shared-config my-website my-service
+```
+
+This'll leave you with three folders to work with, one for each stack:
+
+```bash
+$ tree .
+.
+├── my-service
+├── my-website
+└── shared-config
+```
+
+Change to the `shared-config` folder and generate a new [Pulumi YAML]({{< relref "/docs/intro/languages/yaml" >}}) project, accepting the defaults to create a new `dev` stack:
+
+```bash
+$ cd shared-config
+$ pulumi new aws-yaml
+```
+
+When the new-project wizard completes, add two new settings for the `dev` stack with [`pulumi config set`]({{< relref "/docs/reference/cli/pulumi_config_set" >}}):
+
+```bash
+$ pulumi config set motd 'Hello from Pulumi!'
+$ pulumi config set motd_secret 'Ssh! This is a secret.' --secret
+```
+
+These are the two values we're going to make available to Stacks 2 and 3.
+
+Now let's build out the Pulumi program to provision these items as parameters in AWS Systems Manager. Replace the contents of `Pulumi.yaml` with the following YAML program, which reads the config values we just set, declares two new `aws.ssm.Parameter` resources (using [fully-qualified paths](https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-paramstore-hierarchies.html) as a best practice), and exports their parameter names as Pulumi stack outputs:
+
+```yaml
+name: shared-config
+runtime: yaml
+
+# The values we're reading from stack configuration.
+configuration:
+  motd:
+    type: String
+  motd_secret:
+    type: String
+
+# The AWS SSM Parameters managed by this stack.
+resources:
+  motd_param:
+    type: aws:ssm:Parameter
+    properties:
+      name: /shared-config/dev/motd
+      value: ${motd}
+      type: String
+  motd_secret_param:
+    type: aws:ssm:Parameter
+    properties:
+      name: /shared-config/dev/motd_secret
+      value: ${motd_secret}
+      type: SecureString
+
+# Outputs are the fully-qualified AWS SSM Parameter names
+# that consuming stacks can use to retrieve associated values.
+outputs:
+  motd_param_name: ${motd_param.name}
+  motd_secret_param_name: ${motd_secret_param.name}
+```
+
+At this point, you might be wondering why we're only exporting our parameters' _names_, and not their values.
+
+We certainly _could_ export their values, since we already have them at hand. But if we did, there wouldn't be much point to using Parameter Store at all, since Pulumi itself is perfectly capable of managing plain-text and encrypted values, and stack references give other stacks an easy way of accessing them programmatically. Why not export the values, then?
+
+For one, it'd be less flexible. If we exposed these values as stack outputs, downstream stacks referencing them would only be able to read them at deploy-time --- i.e., _during_ their own Pulumi updates. For `shared-config` values that change rarely, that might be okay, but for others, it'd be less than ideal in that all downstream stacks would have to be updated immediately whenever one of those values were changed.
```

Comment:
You kind of mention it, but maybe not explicitly enough - but only exporting the name of the parameters actually allows to better RBAC/ABAC restrictions on which downstream stacks can actually read the values too.

Reply:
Ah, nice point! I'll see if I can work that in.